Upgrade to importlib and pyproject.toml

igordejanovic opened the pull request:
Hi @goto40. There have been recent reports about setuptools `pkg_resources` API deprecation and complete removal from Python 3.12. This PR started as a work on migrating to `importlib.metadata` but I took an opportunity to do some other cleanups I've planed to do for the next release which will be mostly maintenance.

Arpeggio is upgraded accordingly to CI to pass but the release of Arpeggio will have to wait as there are some important rework to be done for the next release. Nevertheless, we could release textX as these changes work with the current release of Arpeggio.

Here is an excerpt from the CHANGELOG of what is done in this PR:
- **(BIC)** Removed Python support for 3.6 and 3.7. The minimal supported version is 3.8.
- Migrated to pyproject.toml for project configuration.
- Migrated to importlib for language/generator registration/discovery as
  setuptools pkg_resources is deprecated and unsupported from Python 3.12. See
  #411.
- Use [ruff] instead of flake8 for linting. ruff can also be used for code
  formating.
- Relaxed dependency for click to accept 8.x. See #414.
- Use [flit] for package building and publishing.

 [ruff]: https://github.com/astral-sh/ruff
 [flit]: https://flit.pypa.io/

igordejanovic committed:
upgrade to pyproject.toml and importlib

pkg_resources is deprecated and removed from Python 3.12.

closes #411

igordejanovic committed:
remove unused Travis config

igordejanovic committed:
use flit for package build

igordejanovic committed:
add old regression test

igordejanovic committed:
fix code with ruff

igordejanovic committed:
reformat code with ruff

igordejanovic committed:
replace flake8 with ruff

igordejanovic committed:
remove Python 2.x unicode compatibility

igordejanovic committed:
add ruff linter fixes

igordejanovic committed:
drop Python 3.6/3.7 support, add 3.12

- importlib supports only 3.8+
- ruff supports only 3.7+

igordejanovic committed:
add importlib_metadata for backward comp.

igordejanovic committed:
reformat with ruff

igordejanovic committed:
update contributing guide

igordejanovic committed:
relax valid click versions

close #414

igordejanovic committed:
update CHANGELOG

igordejanovic committed:
fix CI script

goto40 commented on pyproject.toml:
```diff
@@ -0,0 +1,116 @@
+    "F",
+    # pyupgrade
+    "UP",
+    # flake8-bugbear
```
flake8: by purpose (flake removed??) ?

igordejanovic replied:
It's just a set of rules ruff use. flake is removed.

goto40 commented on pyproject.toml:
```diff
@@ -0,0 +1,116 @@
+    "UP",
+    # flake8-bugbear
+    "B",
+    # flake8-simplify
```
flake8: by purpose?

goto40 commented on textx/model_params.py:
```diff
@@ -3,12 +3,13 @@
 
-if sys.version < '3':
-    from collections import Mapping
+if sys.version < "3":
```
we do not support python <3: remove completely?

igordejanovic replied:
Good catch. Yeah, we don't need these kind of check anymore. Interesting enough this guard was completely useless as both branches of the if statement were the same.

goto40 commented on textx/registration.py:
```diff
@@ -1,14 +1,18 @@
-import pkg_resources
+import sys
+
+if sys.version_info < (3, 10):
```
same code in if/elseif:  remove completely?

igordejanovic replied:
We still need this to differentiate between 3.{8.9} and newer Python versions as the importlib is API has been changed between versions. In addition for these older version we need additional backward compatibility dependency. You can see that in the `pyproject.toml`. I'll add this note to the changelog.

goto40 commented on textx/cli/__init__.py:
```diff
@@ -1,17 +1,24 @@
+
+import sys
+
+if sys.version_info < (3, 10):
```
same code in if/elseif:  remove completely?

igordejanovic replied:
Same here.

goto40 commented:
found `__future__` in
* `examples/Entity/entity_test.py`
* `tests/perf/test_speed.py`

--> remove?

goto40 commented:
Found `(object)` in *.md and also some *.py files (17 cases in 13 files alltogether)... You changed many of such cases. Maybe also change these 17 places as well?

```
      1 ./docs/tutorials/robot.md:    class Robot(object):
      2 ./docs/tutorials/entity/srcgen.py:class SimpleType(object):
      3 ./docs/tutorials/entity.md:    class SimpleType(object):
      4 ./docs/tutorials/entity.md:    class SimpleType(object):
      5 ./docs/metamodel.md:    class Entity(object):
      6 ./docs/metamodel.md:    class Entity(object):
      7 ./textX.egg-info/PKG-INFO:        class Point(object):
      8 ./textx/metamodel.py:            obj(object): A python object to set attributes to.
      9 ./README.md:class Point(object):
     10 ./examples/drawing/drawing.py:    class Point(object):
     11 ./examples/StateMachine/state_machine.py:class StateMachine(object):
     12 ./examples/robot/robot.py:class Robot(object):
     13 ./examples/Entity/entity_test.py:class SimpleType(object):
     14 ./examples/expression/boolean.py:class Bool(object):
     15 ./examples/expression/boolean.py:class ExpressionElement(object):
     16 ./examples/expression/calc.py:class Calc(object):
     17 ./examples/expression/calc.py:class ExpressionElement(object):
```

goto40 commented:
I found that in august we had 338 tests (https://github.com/textX/textX/actions/runs/5923292736/job/16058653693), now we have 339 tests (https://github.com/textX/textX/actions/runs/6812715815/job/18525682009?pr=416) for python 3.10.

So you added one test?

goto40 commented on tests/functional/test_metamodel/test_multi_metamodel_refs.py:
```diff
@@ -29,24 +32,22 @@
+    class A:
         def __init__(self, **kwargs):
-            super(A, self).__init__()
+            super().__init__()
```
I found some places where no such refactoring took place:
```
      1 ./textx/model.py:442:                return super(user_class, obj).__getattribute__(name)
      2 ./textx/model.py:451:                        return super(user_class, obj).__setattr__(name, value)
      3 ./textx/model.py:460:                        return super(user_class, obj).__delattr__(name)
      4 ./examples/expression/boolean.py:41:        super(ExpressionElement, self).__init__()
      5 ./examples/expression/boolean.py:65:        super(Not, self).__init__(**kwargs)
      6 ./examples/expression/calc.py:43:        super(ExpressionElement, self).__init__()
      7 ./examples/expression/calc.py:49:        super(Factor, self).__init__(**kwargs)
```

igordejanovic replied:
Fixed. `super(user_class, obj)` is `model.py` is still needed.

igordejanovic committed:
linter fixes in examples and perf tests

igordejanovic committed:
fix docs per codereview comment

https://github.com/textX/textX/pull/416#issuecomment-1804121076

goto40 commented:
@igordejanovic : I am not yet finished.... Give me this evening,,,

igordejanovic committed:
fix CI coveralls failure

igordejanovic commented:
> found `__future__` in
> 
>     * `examples/Entity/entity_test.py`
> 
>     * `tests/perf/test_speed.py`
> 
> 
> --> remove?
> Found (object) in *.md and also some *.py files (17 cases in 13 files alltogether)... You changed many of such cases. Maybe also change these 17 places as well?

Ah, yes. I forgot to run ruff over examples and performance tests. Done that now. Fixed other issues as well and added examples and whole tests folder to linter cheks in `runtests.sh`.

> I found that in august we had 338 tests (https://github.com/textX/textX/actions/runs/5923292736/job/16058653693), now we have 339 tests (https://github.com/textX/textX/actions/runs/6812715815/job/18525682009?pr=416) for python 3.10.
> 
> So you added one test?

Yes, good catch :). I added one test for an old issue which I was not commited and I found it lingering in my local repo. Commit 9feddfd73cf0d1fd5cfacc334486a24902500948.

> @igordejanovic : I am not yet finished.... Give me this evening,,,

Sure, take your time.

goto40 commented on tests/functional/test_user_classes.py:
```diff
@@ -56,55 +57,58 @@ def __init__(self, parent, sec):
     # Test that user class is instantiated
     assert type(model).__name__ == "First"
-    assert type(model) is First
+    assert isinstance(model, First)
```
what is the reason to prefer `isinstance` here... In many other locations this was not changed... (?) This ok for me. We do no bother with base classes here. ...just a note/question.

igordejanovic replied:
ruff reports all usages of `type(x) is y`. In most cases it is better to use `isinstance` but in this particular case we don't have to, so we could either change to `isinstance` or use `# noqa` to silence the linter.

goto40 commented on textx/scoping/rrel.py:
```diff
@@ -368,59 +393,67 @@ def get_from_zero_or_more(obj, lookup_list, matched_path,
-            yield iobj, ilookup_list, imatched_path
+        yield from intern_get_next_matches(
+            obj, lookup_list, allowed, matched_path, first_element, 0
+        )
```
If this was done by an automated refactoring; wow!

goto40 replied:
I believe in unittests ;-)

igordejanovic replied:
> If this was done by an automated refactoring; wow!

I have to disappoint you but no, refactor of this piece is manual but ruff nicely reported the issue. :smile:
ruff is mainly linter/formater but it can do some light refactorings like sorting imports, refactor `str.format` to f-strings etc. which is nice especially when dealing with larger code with a lot of instances of these simple refactorings.

goto40 reviewed (approved):
My overall impression is ok! Good work, @igordejanovic.

I appreciate the automated formatting support we get with the new tool!

I thing we have a good unittest coverage and, thus, I do not see any problems with such large scale refactorings/reformatting.

Good work!

igordejanovic committed:
remove unnecessary version guard

igordejanovic committed:
update CHANGELOG

igordejanovic committed:
add ruff to dev dependencies

igordejanovic committed:
improve contributing guide

igordejanovic commented:
> My overall impression is ok! Good work, @igordejanovic.
> 
> I appreciate the automated formatting support we get with the new tool!
> 
> I thing we have a good unittest coverage and, thus, I do not see any problems with such large scale refactorings/reformatting.
> 
> Good work!

Excellent. We are good to go then. I did some tidying of the changelog and will be releasing the new major version during the day. This will be a maintenance release so there is really no changes for folks who are using Python 3.8+.

igordejanovic committed:
fix linter issue
